lucene/core/src/java/org/apache/lucene/search/ParallelSortedCollector.java

```diff
@@ -0,0 +1,612 @@
+    private int numExcludedBound;
+
+    /** A lower bound on the total hits collected by all leaves */
+    private volatile int totalCollected;
+
+    /**
+     * @param numHits the number of hits to be returned
+     * @param collectionThreshold collect at least this many, even if numHits is less, for the sake of counting
+     * @param numThreads the number of threads that will be updating this concurrently. Controls the update interval.
+     *                   If null, an internal default is used.
+     */
+    MaxScoreTerminator(int numHits, int collectionThreshold, Integer numThreads) {
+      this.numHits = numHits;
+      this.totalToCollect = Math.max(numHits, collectionThreshold);
+      thresholdState = new LeafState();
+      thresholdState.score = Double.MAX_VALUE;
+      thresholdState.docid = Integer.MAX_VALUE;
+      if (numThreads != null) {
+        int numThreadsLog2 = 31 - Integer.numberOfLeadingZeros(numThreads);
+        setIntervalBits(numThreadsLog2 + 1);
+      } else {
+        setIntervalBits(DEFAULT_INTERVAL_BITS);
+      }
+    }
+
+    /**
+     * @param sort the query sort
+     * @param numHits the number of hits requested
+     * @return whether the Sort is compatible with early termination using {@link MaxScoreTerminator} and the number of hits
+     * requested is large enough to make this optimization likely to help.
+     */
+    static boolean isApplicable(Sort sort, int numHits) {
+      return numHits > TERMINATION_STRATEGY_HIT_THRESHOLD && sortIsNumeric(sort);
+    }
+
+    private static boolean sortIsNumeric(Sort sort) {
+      switch (sort.getSort()[0].getType()) {
+        case DOC:
+        case INT:
+        case FLOAT:
+        case LONG:
+        case DOUBLE:
+          return true;
+        default:
+          return false;
+      }
+    }
+
+    synchronized LeafState addLeafState() {
+      LeafState newLeafState = new LeafState();
+      newLeafState.index = leafStates.size();
+      leafStates.add(newLeafState);
+      scratch.add(new LeafState());
+      return newLeafState;
+    }
+
+    /**
+     * Set the update interval to 2^bitCount and the intervalMask to 2^bitCount-1. This controls the
+     * rate at which multiple threads report their worst scores. For best performance this should be
+     * set to the nearest power of 2 &gt; the number of expected calling threads.
+     * @param bitCount the number of bits in the interval/intervalMask
+     */
+    void setIntervalBits(int bitCount) {
+      interval = 1 << bitCount;
+      intervalMask = interval - 1;
+    }
+
+    /**
+     * Must be called by leaf collectors on every interval hits to update their progress; they should
+     * call when ((leafCollector.numCollected &amp; this.intervalMask) == 0).
+     * @param newLeafState the leaf collector's current lowest score
+     * @return whether the collector should terminate
+     */
+    synchronized boolean updateLeafState(LeafState newLeafState) {
```

Comment:
Updating the max score should be synchronized but I don't think we should check termination under a lock. It should be ok to check an outdated max score as long as checks are  frequents.

Reply:
hmm, I'm not sure - see my comment above about the need to sort the leaves (or at least find the worst leaf after terminating another).